apps/vmq_diversity/src/vmq_diversity_plugin.erl

```diff
@@ -200,21 +200,47 @@ auth_on_register(Peer, SubscriberId, UserName, Password, CleanSession) ->
 
 auth_on_publish(UserName, SubscriberId, QoS, Topic, Payload, IsRetain) ->
     {MP, ClientId} = subscriber_id(SubscriberId),
-    all_till_ok(auth_on_publish, [{username, nilify(UserName)},
-                                  {mountpoint, MP},
-                                  {client_id, ClientId},
-                                  {qos, QoS},
-                                  {topic, unword(Topic)},
-                                  {payload, Payload},
-                                  {retain, IsRetain}]).
+    case vmq_diversity_cache:match_publish_acl(MP, ClientId, Topic) of
```

Comment:
I think qos and retain should also be part of the match. Otherwise you can't prevent the client from publishing retained messages when using the cache, even if the client is not allowed to do so by the hook logic. Also, I think it should be possible to allow only for specific QoS levels. Of course this is the way I did it in vmq_webhooks and it needs to be optimized so QoS levels and retained bits get's merged into one entry to prevent combinatorial cache size explosions.